src/php/lib/Grpc/Interceptor.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+namespace Grpc;
+
+/**
+ * Represents an interceptor that intercept operations for 4 methods before call starts.
+ */
+class Interceptor
+{
+    public function interceptUnaryUnary(
+      $method,
+      $argument,
+      $deserialize,
+                             array $metadata = [],
+      array $options = [],
+      $continuation
+  ) {
+        return $continuation($method, $argument, $deserialize, $metadata, $options);
+    }
+
+    public function interceptStreamUnary(
+      $method,
+      $deserialize,
+      array $metadata = [],
+                              array $options = [],
+      $continuation
+  ) {
+        return $continuation($method, $deserialize, $metadata, $options);
+    }
+
+    public function interceptUnaryStream(
+      $method,
+      $argument,
+      $deserialize,
+                              array $metadata = [],
+      array $options = [],
+      $continuation
+  ) {
+        return $continuation($method, $argument, $deserialize, $metadata, $options);
+    }
+
+    public function interceptStreamStream(
+      $method,
+      $deserialize,
+                               array $metadata = [],
+      array $options = [],
+      $continuation
+  ) {
+        return $continuation($method, $deserialize, $metadata, $options);
+    }
+
+    public static function intercept($channel, $interceptors)
+    {
+        if (is_array($interceptors)) {
+            for ($i = count($interceptors) - 1; $i >= 0; $i--) {
+                $channel = new InterceptorChannel($channel, $interceptors[$i]);
+            }
+        } else {
+            $channel =  new InterceptorChannel($channel, $interceptors);
+        }
+        return $channel;
+    }
+}
+
+class InterceptorChannel
```

Comment:
Is this class private by default? (Making it private is a good idea, as it decreases our API compatibility surface).

Reply:
It's public. Not sure how to write private class. I'll try.

Reply:
Seems not supported

Reply:
Are all our classes exposed? If not, can we put it in an internal folder designating its private-ness? @stanley-cheung, how do you think we should go about making a class private (does PHP support private inner classes? That might be a good solution.)

Reply:
Didn't find any useful tools. Anonymous classes is supported after PHP 7 but grpc should support PHP5.
Except this. Is this PR ready to be pushed to the grpc repo?
I'll add more comments to all function defined before create PR. Although the format looks strange using official php-cs-format tools, I'll verify some of features later.

Reply:
I'd say yes, it is ready to be pushed as a PR (but not merged) to the grpc repo. Let's squash it first though.

Before we merge *that* PR, we need to work on a proposal and get it through the gRFC process.